### DISCUSSION TITLE
Hosting redash under non-root for newer redash version

### DISCUSSION THREAD

**Discussion by surajsrivathsa** [getredash/redash#6644] [Q&A] [answered]
Is there any way where we can host redash under non-root path for example under `/redash_app/container_1`. 

i tried following two ways of doing this, but unfortunately both of them are outdated and does not work for redash v7/v8 as they first server js which then makes further calls. 

Is there any process that i can follow to change the code or others to host redash under non-root url for newer versions.

Tried Solutions:

1. [nginx reverse proxy + response rewrite](https://discuss.redash.io/t/hosting-redash-not-under-root/135): The rewrite paths do not match with current versions and furthermore js makes more calls which is not working. I also modified response bodies to match the prefix but the internal api calls wont allow.
2. [changing redash code](https://github.com/getredash/redash/issues/1278) : I tried changing application_root, updating templates and making api calls relative but it gets stuck at login. Even public dashboards are not served and making direct api calls with http://localhost:5000/api/**/config or session does not work

**Comment by surajsrivathsa** [comment, marked as answer]
ok some minor positive news. 

I was able to somehow change the code of redash server to be hosted on a non-root path. however this is still incomplete. I am facing issues in displaying list of dashboards/queries/creating dashboard and others. This is on Redash Version: 8.0.2+b37747 (a9d7ca43). I think the below is neither the optimal approach but atleast i was able to see public dashboards and interact with them. 

Please let me know if there are any changes i can make to get other functionalities working and also optimized way of making code change.

Changes made.

- [ ] Flask application

1. `redash.handlers.base.py` ==> Change routes variable and add url prefix. For example `routes = Blueprint('redash', __name__, template_folder=settings.fix_assets_path('templates'), url_prefix="/non_root/prefix")`
2. `redash.authentication.__init__.py` ===> make the url external as true in function `logout_and_redirect_to_index` for example: `index_url = url_for('redash.index', _external=False)` .
3. `redash.handlers.api.py`  ===> add your url prefix `/non_root/prefix` to apis mentioned. For example: `api.add_org_resource(PublicDashboardResource, '/non_root/prefix/api/dashboards/public/<token>', endpoint='public_dashboard')`.
4. Test the apis with postman and check if you are getting same response with and without api. I tested config/session/public_dashboard apis.

- [ ] Angular App

Angular app
1. Search for `'/api'` and make them ` 'api/'` relative in all relevant files. i remember one with ` '/api/config' ` and others which were made relative
2. Either rebuild angular with new base href or simply add base href parameter in `client.dist.index.html` . I added this directly onto built angular app as temporary fix. You need to add this line before title tag in index.html For example: `<base href="/non_root/prefix/">`.
3. Tried logging in and out in a guest browser. Tried also accessing public dashboards which worked. Dashboards and queries were not listed after login which has to be fixed.

**Reply by surajsrivathsa** [reply]
Ok further updates I was able almost get every feature running on non-root path but i made changes inside dist folder and could not rebuild

**Reply by anemology** [reply]
@surajsrivathsa thanks for sharing! You could check my comment in #6319, but the Redash version is not the same as yours 😂

**Reply by surajsrivathsa** [reply]
@anemology : sure i will check it out
